numpy/lib/function_base.py

```diff
@@ -3792,16 +4027,135 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
                 else:
                     r = a.dtype.type(np.nan)
             else:
-                r[..., n.squeeze(0)] = a.dtype.type(np.nan)
+                r[n.squeeze(axis)] = a.dtype.type(np.nan)
         else:
             if r.ndim == 1:
                 r[:] = a.dtype.type(np.nan)
             else:
-                r[..., n.repeat(q.size, 0)] = a.dtype.type(np.nan)
+                r[..., n.squeeze(axis)] = a.dtype.type(np.nan)
 
     return r
 
 
+def _indices_for_weighted_quantile(ap, q, axis, weights):
+    """
+    Computes the indices of weighted quantiles along axis -1.
+
+    Parameters
+    ----------
+    ap : array_like
+        Array where weighted quantiles are evaluated along axis.
+    q : array_like
+        Quantiles sought.
+    axis : int
+        Axis along which to evaluate quantiles.  Will be moved to -1 for
+        np.vectorize operations.
+    weights : array_like
+        An array of weights associated with the values in `ap`.
+
+    Returns
+    -------
+    ap : array_like
+        An expanded twice as large as the input `ap` array to perform
+        interpolation on.  The axis of interest is moved to -1.
+    indices : array_like
+        Indices along the -1 axis of ap to interpolate.  All floats.
+    Nx : scalar
+        Length of the axis of ap.
+    """
+    # first move to axis -1 for np.vectorize
+    ap = np.swapaxes(ap.astype('f8'), axis, -1)
+    weights = np.swapaxes(weights.astype('f8'), axis, -1)
+
+    # values with weight=0 are made nan and later sorted to end of array
+    if ap[weights == 0.].any():
+        ap[weights == 0.] = np.nan
+
+    def _sort_by_index(vector, vec_indices):
+        return vector[vec_indices]
+    # this func vectorizes sort along axis
+    arraysort = np.vectorize(_sort_by_index, signature='(i),(i)->(i)')
+
+    # compute the sorted data array.  NaNs are sorted to the end of array
+    ind_sorted = np.argsort(ap, axis=-1)  # sort values long axis
+    ap_sorted = arraysort(ap, ind_sorted)
+
+    # align the weights to the sorted data array
+    ws_sorted = arraysort(weights, ind_sorted)
+
+    # along axis -1,
+    # 1.) normalize weights by the smallest non-zero weight
+    # 2.) if more than 1 weights, subtract the first one by unit weight
+    def _normalize_and_set_boundary(ws):  # ws is a 1-D vector
+        w_vec = ws.copy()
+        inds = w_vec > 0.
+        if (w_vec[inds] < 1.0).any():
+            w_vec[inds] = w_vec[inds] / w_vec[inds].min()
```

Comment:
I don't understand the purpose of this. Why bother filtering to `w_vec > 0`, if that's forbidden anyway?

Reply:
A weight is allowed to be 0.  The constraint on weights are

1.) >= 0
2.) cannot sum to 0

When a weight is 0, it just means that the corresponding cell in `ap` is removed from further consideration (hence setting it to `np.nan` earlier).  Here I'm normalizing the weights by the smallest value, and I want to ensure that smallest value is not  0, or I'd get a `np.nan` everywhere.